### PR TITLE
Use custom filter generator for course discovery search

### PR DESCRIPTION
**Pull request by kluo**
@stvstnfrd

### CODE REVIEW

File: lms/lib/courseware_search/lms_filter_generator.py [hunk 1]
```diff
@@ -28,11 +28,10 @@ def _enrollments_for_user(self, user):
     def field_dictionary(self, **kwargs):
         """ add course if provided otherwise add courses in which the user is enrolled in """
         field_dictionary = super(LmsSearchFilterGenerator, self).field_dictionary(**kwargs)
-        if not kwargs.get('user'):
-            field_dictionary['course'] = []
-        elif not kwargs.get('course_id'):
-            user_enrollments = self._enrollments_for_user(kwargs['user'])
-            field_dictionary['course'] = [unicode(enrollment.course_id) for enrollment in user_enrollments]
+        if kwargs.get('user'):
```
Comment (stvstnfrd) [inline lms/lib/courseware_search/lms_filter_generator.py]: Why these changes?

What are we accomplishing?

Why do we need to modify the default filter if we're implementing our own (that subclasses it)?

File: openedx/stanford/lms/envs/common.py [hunk 2]
```diff
@@ -114,6 +114,7 @@
 # The following fields are available in the URL: {course_id} {student_id}
 PROGRESS_SUCCESS_BUTTON_URL = 'http://<domain>/<path>/{course_id}'
 PROGRESS_SUCCESS_BUTTON_TEXT_OVERRIDE = None
+SEARCH_FILTER_GENERATOR = 'openedx.stanford.lms.lib.courseware_search.lms_filter_generator.TileSearchFilterGenerator'
```
Comment (stvstnfrd) [inline openedx/stanford/lms/envs/common.py]: Yeah, I'm okay with leaving this here for now;
we can move to JSON when the need arises.

File: openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+using course tiles for discovery search
+"""
+from search.filter_generator import SearchFilterGenerator
+
+from branding_stanford.models import TileConfiguration
+from lms.lib.courseware_search.lms_filter_generator import LmsSearchFilterGenerator
+
+
+class TileSearchFilterGenerator(LmsSearchFilterGenerator):
+    """ SearchFilterGenerator for LMS Search """
```
Comment (stvstnfrd) [inline openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py]: https://github.com/Stanford-Online/edx-platform/wiki/docstrings

File: openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    def field_dictionary(self, **kwargs):
+        """ Return field filter dictionary for search """
+        # Adds course filter if provided, for in course search
+        field_dictionary = SearchFilterGenerator.field_dictionary(self, **kwargs)
```
Comment (stvstnfrd) [inline openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py]: You mentioned that this is lifted from the LMS default,
but I see that we now omit the call to `super`.
- Was this an explicit omission?
- What is the impact/result of this change?

I'm wondering if we shouldn't do the same thing they do in the default;
start by grabbing the result of the `super` call,
then basically post-process it after w/ our customization.

Generally speaking, if a parent class also already calls up to the grandparent's method,
you generally want to also have the child call up to the parents method,
as the GP call indicates that the children depend on some logic having been executed "upstream".

File: openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        if kwargs.get('user'):
+            if not kwargs.get('course_id'):
```
Comment (stvstnfrd) [inline openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py]: What happens if there _is_ a `user` _and_ there _is_ a `course_id`?

What is the fall-through logic?

File: lms/lib/courseware_search/lms_filter_generator.py (hunk 1 above)
Comment (kluo) [reply lms/lib/courseware_search/lms_filter_generator.py]: Returning an empty course list when there's no user interacted with the former [hardcoded search api logic](https://github.com/edx/edx-search/blob/master/search/api.py#L89) such that it ends up just being an empty dictionary, i.e no filter, so it would display all courses. So that was the default behavior. Now that I had to remove the hardcoded `org` filter, if this logic remained, it would then filter on an empty course list and instead return nothing. To maintain the prior default behavior, I updated this logic

File: openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py (hunk 4 above)
Comment (kluo) [reply openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py]: Yes I should clarify I only subclasses the default `LmsSearchFilterGenerator` so I wouldn't have to duplicate the enrollment code https://github.com/Stanford-Online/edx-platform/blob/master/lms/lib/courseware_search/lms_filter_generator.py#L22
but since we don't care about the org filter logic that it has I called the base class `field_dictionary` instead.

File: openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py (hunk 5 above)
Comment (kluo) [reply openedx/stanford/lms/lib/courseware_search/lms_filter_generator.py]: The base class handles if there is a `course_id` specified, which means its an in-course search: https://github.com/edx/edx-search/blob/master/search/filter_generator.py#L27
